repobee_sanitizer/_fileutils.py

```diff
@@ -0,0 +1,87 @@
+"""Functions for dealing with the file system.
+
+.. module:: _fileutils
+    :synopsis: Functions for dealing with the file system.
+"""
+import subprocess
+import pathlib
+import collections
+import sys
+
+
+class RelativePath(
+    collections.namedtuple("RelativePath", "relpath encoding".split())
+):
+    """Data structure containing a relative path, and the file encoding of the
+    file from which the relative path was created. The read and write methods
+    must be used relative to some base directory.
+    """
+
+    def read_text_relative_to(self, basedir: pathlib.Path) -> str:
+        """Read text from this relative path's file with the expected encoding,
+        rooted in basedir.
+
+        Args:
+            basedir: The base directory which the path is relative to.
+        Returns:
+            The content of the file this path points to relative to the base
+            directory, using the encoding of the original file.
+        """
+        return (basedir / self.relpath).read_text(encoding=self.encoding)
+
+    def write_text_relative_to(self, data: str, basedir: pathlib.Path) -> None:
+        """Write the provided string to the file at this path's location
+        relative to the basedir, using the encoding of the original file this
+        path was created from.
+
+        Args:
+            data: Data to write.
+            basedir: The base directory to which this path is relative.
+        """
+        (basedir / self.relpath).write_text(data=data, encoding=self.encoding)
+
+    @property
```

Comment:
Couldn't find a concise explanation of this online

Reply:
that is the @property I'm referring to

Reply:
The short version is that it allows us to use the annotated method like it's an attribute of the object, rather than a method, i.e. we can call the method without supplying parentheses. Here's an example:

```python
class A:
    def value(self):
        return 42

class B:
    @property
    def value(self):
        return 42

# both a_val and b_val will get the value 42
a_val = A().value()
b_val = B().value
```

[This RealPython article](https://realpython.com/python-descriptors/) goes into much more detail if you're interested.